File: up/ina219.py
from ustruct import pack, unpack

class Ina219(object):
    def __init__(self, i2c, adr=0x40, r_shunt=50):
        """
        voltage / current / power monitor
        r_shunt [mOhm]
        """
        self.i2c = i2c
        self.adr = adr
        self._w(0, 1 << 15)  # reset
        self.cfg = self._r(0)
        self.r_shunt = r_shunt

    def _r(self, reg, signed=False):
        val_ = self.i2c.readfrom_mem(self.adr, reg, 2)
        return unpack('>h' if signed else '>H', val_)[0]

    def _w(self, reg, val):
        ''' always unsigned '''
        val_ = int.to_bytes(val, 2, 'big')
        self.i2c.writeto_mem(self.adr, reg, val_)

    def _setcfg(self, shift, n_bits, val, write=True):
        mask = (1 << n_bits) - 1
        mask <<= shift
        val <<= shift
        self.cfg &= ~mask
        self.cfg |= (val & mask)
        if write:
            self._w(0, self.cfg)

    def off(self):
        self._w(0, self.cfg & 0xFFF8)

    @property
    def bus_32v(self):
        return (self.cfg >> 13) & 0x1

    @bus_32v.setter
    def bus_32v(self, val):
        '''
        bus_32v = True (16 V FSR), False (32 V FSR)
        '''
        self._setcfg(13, 1, val)

    @property
    def pga(self):
        return (self.cfg >> 11) & 0x3

    @pga.setter
    def pga(self, val):
        '''
        set shunt voltage gain:
        val = 0 (+- 40 mV), 1 (+- 80 mV), 2 (+- 160 mV) or 3 (+- 320 mV)
        '''
        self._setcfg(11, 2, val)

    @property
    def avg(self):
        return (self.cfg >> 3) & 0x7

    @avg.setter
    def avg(self, val):
        '''
        averaging factor: 0 - 7
        applies to BUS and SHUNT
        always 12 bit mode
        '''
        val |= 8
        self._setcfg(3, 4, val)
        self._setcfg(7, 4, val)

    def on(self):
        self._w(0, self.cfg)

    def getI(self):
        ''' current in [mA] '''
        return self._r(1, True) * 10 // self.r_shunt

    def getV(self):
        ''' voltage in [mV] '''
        return (self._r(2) & 0xFFF8) >> 1

File: pio/src/ina219.cpp
#include "Arduino.h"
#include "Wire.h"
#include "ina219.h"

#define I2C_ADDR 0x40
#define R_SHUNT 50000 // [uOhm]

static uint16_t inaCfg = 0;

static void w(uint8_t reg, uint16_t val)
{
    Wire.beginTransmission(I2C_ADDR);
    Wire.write(reg);
    Wire.write(val >> 8);
    Wire.write(val & 0xFF);
    Wire.endTransmission(true);
}

static void setcfg(unsigned shift, unsigned n_bits, unsigned val)
{
    unsigned mask = (1 << n_bits) - 1;
    mask <<= shift;
    val <<= shift;
    inaCfg &= ~mask;
    inaCfg |= (val & mask);
    w(0, inaCfg);
}

static uint16_t rui(uint8_t reg)
{
    uint8_t buf[2];
    Wire.beginTransmission(I2C_ADDR);
    Wire.write(reg);
    Wire.endTransmission(false);  // Generate repeated start with false
    Wire.readTransmission(I2C_ADDR, buf, 2);
    Wire.endTransmission(true);
    return (buf[0] << 8 | buf[1]);
}

static int16_t ri(uint8_t reg) { return (int16_t)rui(reg); }

void inaInit()
{
    w(0, 1 << 15);  // reset
    inaCfg = rui(0);
}

// current in [mA]
int inaI()
{
    int tmp = ri(1);
    return (tmp * 10 * 1000 / R_SHUNT);
}

// bus voltage in [mV]
int inaV()
{
    int tmp = ri(2) & 0xFFF8;
    return (tmp / 2);
}

// averaging factor: 0 - 7
// applies to BUS and SHUNT
// always 12 bit mode
void inaAvg(unsigned val)
{
    val |= 8;
    setcfg(3, 4, val);
    setcfg(7, 4, val);
}

// set bus voltage scaling
// val = false: 16 V FSR, true: 32 V FSR
void inaBus32(bool val) { setcfg(13, 1, val); }

// set shunt voltage divider
// val = 0: +- 40 mV FSR, 1: +- 80 mV FSR, 2: +- 160 mV FSR, 3: +- 320 mV FSR
void inaPga(unsigned val) { setcfg(11, 2, val); }

// enable power safe mode
void inaOff() { w(0, inaCfg & 0xFFF8); }

// wake up from power safe mode
void inaOn() { w(0, inaCfg); }
